File: app/src/test/java/nz/compliance/app/imports/GoldPlan.java
package nz.compliance.app.imports;

import nz.compliance.engine.model.Point;

import java.util.List;

/**
 * Hand-authored ground truth for one plan image, deserialized from a sibling {@code <image>.gold.json}
 * and compared against a live {@link PlanExtraction} by {@link ExtractionScorer}. All coordinates are
 * IMAGE PIXELS of the committed fixture (mind any downscaling). {@code scaleMetresPerPixel} null means
 * "expect a null scale"; {@code scoreScale} null/true scores the scale dimension, false skips it.
 */
public record GoldPlan(List<GoldRoom> rooms, List<GoldDoor> doors, Double scaleMetresPerPixel, Boolean scoreScale) {

    public GoldPlan {
        rooms = rooms == null ? List.of() : List.copyOf(rooms);
        doors = doors == null ? List.of() : List.copyOf(doors);
    }

    public record GoldRoom(String label, List<Point> polygonPx) {
        public GoldRoom {
            polygonPx = polygonPx == null ? List.of() : List.copyOf(polygonPx);
        }
    }

    public record GoldDoor(List<Point> positionPx, boolean exit) {
        public GoldDoor {
            positionPx = positionPx == null ? List.of() : List.copyOf(positionPx);
        }
    }
}

File: frontend/src/gold/goldPlan.ts
import type { Point } from '../api/types'

/**
 * Ground truth for one plan image. Mirrors GoldPlan.java exactly: a field renamed here is a gold
 * file that parses into nulls and silently scores nothing.
 *
 * All coordinates are IMAGE PIXELS of the committed fixture, so they must be taken against the
 * image's natural size, not against whatever size the SVG happens to be rendered at.
 */
export interface GoldPlan {
  rooms: GoldRoom[]
  doors: GoldDoor[]
  /** null means "the plan has no resolvable scale, and the extractor is expected to report none". */
  scaleMetresPerPixel: number | null
  /** false skips the scale dimension when scoring, for a plan whose scale you have not established. */
  scoreScale: boolean
}

export interface GoldRoom {
  label: string
  polygonPx: Point[]
}

export interface GoldDoor {
  positionPx: Point[]
  exit: boolean
}

/** Pixel coordinates are whole pixels. Rounding here keeps the exported file readable and diffable. */
const round = (p: Point): Point => ({ x: Math.round(p.x), y: Math.round(p.y) })

export function buildGold(
  rooms: GoldRoom[],
  doors: GoldDoor[],
  scaleMetresPerPixel: number | null,
  scoreScale: boolean,
): GoldPlan {
  return {
    rooms: rooms.map((r) => ({ label: r.label, polygonPx: r.polygonPx.map(round) })),
    doors: doors.map((d) => ({ positionPx: d.positionPx.map(round), exit: d.exit })),
    scaleMetresPerPixel,
    scoreScale,
  }
}

/**
 * Metres per pixel from two clicked points a known real distance apart.
 * Returns null rather than Infinity/NaN when the two points coincide, so a mis-click cannot
 * produce a scale that looks like a number and poisons every area on the plan.
 */
export function metresPerPixel(a: Point, b: Point, knownMetres: number): number | null {
  const px = Math.hypot(a.x - b.x, a.y - b.y)
  if (px <= 0 || !(knownMetres > 0)) return null
  const m = knownMetres / px
  return Number.isFinite(m) && m > 0 ? m : null
}

/** foo.png -> foo.gold.json, matching the sibling-file convention ExtractionScorer relies on. */
export function goldFileName(imageFileName: string): string {
  const base = imageFileName.replace(/\.[^./\\]+$/, '')
  return `${base}.gold.json`
}
